Documentation for Goal conditioning (#5149)
Unity-Technologies/ml-agents: docs/Learning-Environment-Design-Agents.md

    - [RayCast Observation Summary & Best Practices](#raycast-observation-summary--best-practices)
  - [Variable Length Observations](#variable-length-observations)
    - [Variable Length Observation Summary & Best Practices](#variable-length-observation-summary--best-practices)

>   - [Goal Signals](#goal-signals)

Author: ```suggestion
  - [Goal Signal](#goal-signal)
```

 of an entity to the `BufferSensor`.
 - Normalize the entities observations before feeding them into the `BufferSensor`.

### Goal Observations

> It is possible for agents to collect observations that will be treated as "goal".

Reviewer: ```suggestion
It is possible for agents to collect observations that will be treated as "goal."
```

A goal is used to condition the policy of the Agent, meaning that if the goal

> changes, the behavior of the Agent will change as well. Note that this is true

Reviewer: I think we should say that goals change the policy, which is a mapping from (observation -> action)

 of an entity to the `BufferSensor`.
 - Normalize the entities observations before feeding them into the `BufferSensor`.

> ### Goal Signals

Author: ```suggestion
### Goal Signal
```

> It is possible for agents to collect observations that will be treated as "goal."

Reviewer: ```suggestion
It is possible for agents to collect observations that will be treated as a "goal."
```

Reviewer: I think within this paragraph we should also call these goal signals, and not goals.

> It is possible for agents to collect observations that will be treated as "goal signals".

Author: ```suggestion
It is possible for agents to collect observations that will be treated as "goal signal".
```

> A goal signal is used to condition the policy of the Agent, meaning that if the goal

Reviewer 2: ```suggestion
A goal signal is used to condition the policy of the agent, meaning that if the goal
```

changes, the policy (i.e. the mapping from observations to actions) will change
as well. Note that this is true
for any observation since all observations influence the policy of the Agent to
some degree. But by specifying a goal signal explicitly, we can make this conditioning
more important to the agent. This feature can be used in settings where an agent
must learn to solve different tasks that are similar by some aspects because the
agent will learn to reuse learnings from different tasks to generalize better.
In Unity, you can specify that a `VectorSensor` or
a `CameraSensor` is a goal by attaching a `VectorSensorComponent` or a

> `CameraSensorComponent` to the Agent can selecting `Goal` as `Observation Type`.

Author: ```suggestion
`CameraSensorComponent` to the Agent and selecting `Goal Signal` as `Observation Type`.
```

On the trainer side, there are two different ways to condition the policy. This
setting is determined by the
[conditioning_type parameter](Training-Configuration-File.md#common-trainer-configurations).
If set to `hyper` (default) a [HyperNetwork](https://arxiv.org/pdf/1609.09106.pdf)
will be used to generate some of the
weights of the policy using the goal observations as input. Note that using a
HyperNetwork requires a lot of computations, it is recommended to use a smaller
number of hidden units in the policy to alleviate this.

> If set to `none` the goal signals will be considered as regular observations.

Author: ```suggestion
If set to `none` the goal signal will be considered as regular observations.
```
